KafkaBasedCommandConsumerFactoryImplIT fails during CI build

The *KafkaBasedCommandConsumerFactoryImplIT* test has failed during the CI build with the following log output:

```sh
[INFO] Running org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT
15:29:39.171 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:39.184 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:39.205 [vert.x-eventloop-thread-0] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:39.807 [vert.x-eventloop-thread-0] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:43.061 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=0,timestamp=1625153383047,key=myDeviceId,value=cmd_0_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 48]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [48])], isReadOnly = false)}
15:29:43.063 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=1,timestamp=1625153383053,key=myDeviceId,value=cmd_1_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 49]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [49])], isReadOnly = false)}
15:29:43.065 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=2,timestamp=1625153383053,key=myDeviceId,value=cmd_2_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 50]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [50])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=3,timestamp=1625153383053,key=myDeviceId,value=cmd_3_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 51]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [51])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=4,timestamp=1625153383053,key=myDeviceId,value=cmd_4_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 52]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [52])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=5,timestamp=1625153383053,key=myDeviceId,value=cmd_5_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 53]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [53])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=6,timestamp=1625153383053,key=myDeviceId,value=cmd_6_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 54]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [54])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=7,timestamp=1625153383053,key=myDeviceId,value=cmd_7_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 55]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [55])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=8,timestamp=1625153383053,key=myDeviceId,value=cmd_8_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 56]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [56])], isReadOnly = false)}
15:29:43.072 [vert.x-eventloop-thread-0] INFO  o.e.h.t.c.KafkaBasedCommandConsumerFactoryImplIT - received KafkaConsumerRecord{topic=hono.command_internal.myAdapterInstanceId_18fc729b-9a5d-4cb1-97e4-4b465baed64b,partition=0,offset=9,timestamp=1625153383053,key=myDeviceId,value=cmd_9_payload,headers=RecordHeaders(headers = [RecordHeader(key = device_id, value = [109, 121, 68, 101, 118, 105, 99, 101, 73, 100]), RecordHeader(key = subject, value = [99, 109, 100, 95, 57]), RecordHeader(key = tenant_id, value = [116, 101, 110, 97, 110, 116, 95, 98, 54, 49, 48, 54, 50, 102, 57, 45, 55, 102, 99, 99, 45, 52, 97, 53, 48, 45, 56, 55, 100, 97, 45, 100, 54, 98, 98, 52, 100, 48, 48, 101, 49, 54, 102]), RecordHeader(key = orig-partition, value = [48]), RecordHeader(key = orig-offset, value = [57])], isReadOnly = false)}
15:29:43.123 [vert.x-eventloop-thread-2] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:43.125 [vert.x-eventloop-thread-2] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:45.563 [kafka-producer-network-thread | generic-sender-3c33b420-742c-4e65-8fec-5d190e7bf460] WARN  o.apache.kafka.clients.NetworkClient - [Producer clientId=generic-sender-3c33b420-742c-4e65-8fec-5d190e7bf460] Error while fetching metadata with correlation id 22 : {hono.command.f384d7f5-ee47-4782-bf25-ebae679046c3=LEADER_NOT_AVAILABLE, hono.command.42c73be2-e8a6-4a08-bd64-c995e9f576ad=LEADER_NOT_AVAILABLE, hono.command.4c5da317-a8a1-411a-847c-2ba4a7be5303=LEADER_NOT_AVAILABLE, hono.command.0fc2f83e-2125-4e4b-b005-61b78d3828ca=LEADER_NOT_AVAILABLE, hono.command.851b3803-9054-41ae-ba88-b3f277ec0405=LEADER_NOT_AVAILABLE, hono.command.63cb6acd-70ba-422f-b2a2-9d985422af93=LEADER_NOT_AVAILABLE, hono.command.395aacd9-6b6a-446d-b5e9-a02e0cc2e6dd=LEADER_NOT_AVAILABLE, hono.command.19fa9954-89db-4f46-b4e7-c246eb29802c=LEADER_NOT_AVAILABLE, hono.command.8b6467b9-0943-4f30-9bb0-1e28408102a0=LEADER_NOT_AVAILABLE}
CMDROUTER15:29:45.623 [kafka-producer-network-thread | cmd-router-command_response-8f3294f0-7f48-476c-bb97-2a96781502e6] WARN  o.apache.kafka.clients.NetworkClient - [Producer clientId=cmd-router-command_response-8f3294f0-7f48-476c-bb97-2a96781502e6] Error while fetching metadata with correlation id 27 : {hono.command_response.63cb6acd-70ba-422f-b2a2-9d985422af93=LEADER_NOT_AVAILABLE, hono.command_response.851b3803-9054-41ae-ba88-b3f277ec0405=LEADER_NOT_AVAILABLE, hono.command_response.395aacd9-6b6a-446d-b5e9-a02e0cc2e6dd=LEADER_NOT_AVAILABLE, hono.command_response.466014c2-4821-4a3c-bf52-83bdec62049d=LEADER_NOT_AVAILABLE, hono.command_response.19fa9954-89db-4f46-b4e7-c246eb29802c=LEADER_NOT_AVAILABLE, hono.command_response.42c73be2-e8a6-4a08-bd64-c995e9f576ad=LEADER_NOT_AVAILABLE, hono.command_response.bbd78877-5873-48a1-a7c7-0c97ee609e08=LEADER_NOT_AVAILABLE, hono.command_response.4c5da317-a8a1-411a-847c-2ba4a7be5303=LEADER_NOT_AVAILABLE, hono.command_response.8b6467b9-0943-4f30-9bb0-1e28408102a0=LEADER_NOT_AVAILABLE, hono.command_response.f384d7f5-ee47-4782-bf25-ebae679046c3=LEADER_NOT_AVAILABLE, hono.command_response.2138d944-6936-4a2e-8a16-796cc0350a1d=LEADER_NOT_AVAILABLE, hono.command_response.0fc2f83e-2125-4e4b-b005-61b78d3828ca=LEADER_NOT_AVAILABLE}
15:29:46.417 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - partition [hono.command.tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8-0] isn't being handled anymore but its command queue isn't empty! [queue size: 9]
15:29:46.420 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_2, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.426 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_3, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.427 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_4, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.427 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_5, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.427 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_6, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.427 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_7, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.428 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_8, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.428 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_9, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:46.451 [vert.x-eventloop-thread-2] WARN  o.e.h.c.k.c.AsyncHandlingAutoCommitKafkaConsumer - partition [hono.command.tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8-0] not assigned to consumer anymore but latest handled record offset hasn't been committed yet! {lastSequentiallyCompletedOffset=0, lastCommittedOffset=-1, queue=[1, 2 (completed), 3 (completed), 4 (completed), 5 (completed), 6 (completed), 7 (completed), 8 (completed), 9 (completed)]}
15:29:46.454 [vert.x-eventloop-thread-2] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:38041
15:29:46.508 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - commands received from partition [hono.command.tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8-0] aren't handled by this consumer anymore [Command [name: cmd_1, tenant-id: tenant_39ed439d-b4c3-4a0a-af42-ce570e89f2d8, device-id: myDeviceId, request-id: null]]
15:29:56.402 [vert.x-eventloop-thread-2] ERROR io.vertx.core.impl.ContextImpl - Unhandled exception
java.lang.NullPointerException: null
	at org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT.lambda$cleanupAfterTest$4(KafkaBasedCommandConsumerFactoryImplIT.java:148)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:176)
	at io.vertx.core.impl.CompositeFutureImpl.doComplete(CompositeFutureImpl.java:251)
	at io.vertx.core.impl.CompositeFutureImpl.lambda$join$3(CompositeFutureImpl.java:109)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.Future.lambda$map$5(Future.java:438)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:176)
	at io.vertx.core.impl.CompositeFutureImpl.doComplete(CompositeFutureImpl.java:251)
	at io.vertx.core.impl.CompositeFutureImpl.lambda$join$3(CompositeFutureImpl.java:109)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:176)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:21)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:327)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Error:  Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 17.239 s <<< FAILURE! - in org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT
Error:  testCommandsGetForwardedIfOneConsumerInstanceGetsClosed{VertxTestContext}  Time elapsed: 13.265 s  <<< ERROR!
java.util.concurrent.TimeoutException: The test execution timed out. Make sure your asynchronous code includes calls to either VertxTestContext#completeNow(), VertxTestContext#failNow() or Checkpoint#flag()
```

This still seems to happen
```sh
[INFO] Running org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT
15:03:58.250 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:03:58.253 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:03:58.281 [vert.x-eventloop-thread-0] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:03:58.952 [vert.x-eventloop-thread-0] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:04:02.831 [vert.x-eventloop-thread-2] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:04:02.832 [vert.x-eventloop-thread-2] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:04:06.135 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - partition [hono.command.tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40-0] isn't being handled anymore but its command queue isn't empty! [queue size: 9]
15:04:06.136 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_2, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.169 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_3, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.171 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_4, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.172 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_5, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.174 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_6, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.175 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_7, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.176 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_8, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.179 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - its partition isn't being handled anymore [Command [name: cmd_9, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
15:04:06.206 [vert.x-eventloop-thread-2] WARN  o.e.h.c.k.c.AsyncHandlingAutoCommitKafkaConsumer - partition [hono.command.tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40-0] not assigned to consumer anymore but latest handled record offset hasn't been committed yet! {lastSequentiallyCompletedOffset=0, lastCommittedOffset=-1, queue=[1, 2 (completed), 3 (completed), 4 (completed), 5 (completed), 6 (completed), 7 (completed), 8 (completed), 9 (completed)]}
15:04:06.209 [vert.x-eventloop-thread-2] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:35911
15:04:06.260 [vert.x-eventloop-thread-2] INFO  o.e.h.c.i.k.KafkaCommandProcessingQueue - command won't be sent - commands received from partition [hono.command.tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40-0] aren't handled by this consumer anymore [Command [name: cmd_1, tenant-id: tenant_98d8603a-870b-45d7-bb9a-1a0d21a6ed40, device-id: myDeviceId, request-id: null]]
Error:  Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 15.825 s <<< FAILURE! - in org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT
Error:  testCommandsGetForwardedIfOneConsumerInstanceGetsClosed{VertxTestContext}  Time elapsed: 11.258 s  <<< FAILURE!
java.lang.AssertionError: java.lang.AssertionError: did not receive 9 out of 10 expected messages after 8s
	at org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT.testCommandsGetForwardedIfOneConsumerInstanceGetsClosed(KafkaBasedCommandConsumerFactoryImplIT.java:344)

15:04:14.106 [vert.x-eventloop-thread-2] ERROR io.vertx.core.impl.ContextImpl - Unhandled exception
java.lang.NullPointerException: null
	at org.eclipse.hono.tests.commandrouter.KafkaBasedCommandConsumerFactoryImplIT.lambda$cleanupAfterTest$4(KafkaBasedCommandConsumerFactoryImplIT.java:149)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:176)
	at io.vertx.core.impl.CompositeFutureImpl.doComplete(CompositeFutureImpl.java:251)
	at io.vertx.core.impl.CompositeFutureImpl.lambda$join$3(CompositeFutureImpl.java:109)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.Future.lambda$map$5(Future.java:438)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:176)
	at io.vertx.core.impl.CompositeFutureImpl.doComplete(CompositeFutureImpl.java:251)
	at io.vertx.core.impl.CompositeFutureImpl.lambda$join$3(CompositeFutureImpl.java:109)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:176)
	at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:21)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:327)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    KafkaBasedCommandConsumerFactoryImplIT.testCommandsGetForwardedIfOneConsumerInstanceGetsClosed java.lang.AssertionError: did not receive 9 out of 10 expected messages after 8s
[INFO] 
Error:  Tests run: 392, Failures: 1, Errors: 0, Skipped: 8
```

--- a/clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/HonoKafkaConsumer.java
+++ b/clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/HonoKafkaConsumer.java
@@ -82,7 +82,7 @@ public class HonoKafkaConsumer implements Lifecycle {
     protected final AtomicBoolean stopCalled = new AtomicBoolean();
 
     private final Map<String, String> consumerConfig;
-    private final AtomicReference<Promise<Set<TopicPartition>>> subscribeDonePromiseRef = new AtomicReference<>();
+    private final AtomicReference<Promise<Void>> subscribeDonePromiseRef = new AtomicReference<>();
     private final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler;
     private final Set<String> topics;
     private final Pattern topicPattern;
@@ -292,16 +292,6 @@ public class HonoKafkaConsumer implements Lifecycle {
             // the consumer is actually ready to receive records already
             kafkaConsumer.asStream().pollTimeout(Duration.ofMillis(10)); // let polls finish quickly until start() is completed
             subscribeAndWaitForRebalance()
-                    .compose(assignedPartitions -> {
-                        if (!assignedPartitions.isEmpty()
-                                && "latest".equals(consumerConfig.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG))) {
-                            // Wait for offset positions to be reset
-                            // (otherwise records published between start() completion and the offset-reset might get skipped)
-                            // This is done by having KafkaConsumer.updateFetchPositions() be called.
-                            return updateFetchPositions(assignedPartitions);
-                        }
-                        return Future.succeededFuture((Void) null);
-                    })
                     .onSuccess(v2 -> {
                         kafkaConsumer.asStream().pollTimeout(POLL_TIMEOUT);
                         logSubscribedTopicsOnStartComplete();
@@ -311,7 +301,21 @@ public class HonoKafkaConsumer implements Lifecycle {
         return startPromise.future();
     }
 
-    private Future<Void> updateFetchPositions(final Set<TopicPartition> assignedPartitions) {
+    /**
+     * To be invoked after partition assignment, making sure that with the completion of the returned future,
+     * partition offset positions have been set according to either the last committed state or the auto offset
+     * reset config ("latest" being relevant here) if no offset is committed yet.
+     * <p>
+     * This makes sure records published after the returned future is completed are actually received by the consumer.
+     * <p>
+     * Note that the returned future is never failed here.
+     */
+    private Future<Void> ensurePositionsHaveBeenSetIfNeeded(final Set<TopicPartition> assignedPartitions) {
+        // skip if offset reset config set to "earliest", no need to wait for the result here since consumer will receive all records anyway
+        if (assignedPartitions.isEmpty()
+                || "earliest".equals(consumerConfig.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG))) {
+            return Future.succeededFuture();
+        }
         final Promise<Void> resultPromise = Promise.promise();
         runOnKafkaWorkerThread(v -> {
             final Consumer<String, Buffer> wrappedConsumer = kafkaConsumer.asStream().unwrap();
@@ -399,11 +403,11 @@ public class HonoKafkaConsumer implements Lifecycle {
         });
     }
 
-    private Future<Set<TopicPartition>> subscribeAndWaitForRebalance() {
+    private Future<Void> subscribeAndWaitForRebalance() {
         if (stopCalled.get()) {
             return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "already stopped"));
         }
-        final Promise<Set<TopicPartition>> subscribeDonePromise = subscribeDonePromiseRef
+        final Promise<Void> subscribeDonePromise = subscribeDonePromiseRef
                 .updateAndGet(promise -> promise == null ? Promise.promise() : promise);
         final Promise<Void> subscriptionPromise = Promise.promise();
         if (topicPattern != null) {
@@ -437,8 +441,7 @@ public class HonoKafkaConsumer implements Lifecycle {
                 subscribeDonePromise.tryFail(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, errorMsg));
             }
         });
-        return CompositeFuture.all(subscriptionPromise.future(), subscribeDonePromise.future())
-                .map(r -> subscribeDonePromise.future().result());
+        return CompositeFuture.all(subscriptionPromise.future(), subscribeDonePromise.future()).mapEmpty();
     }
 
     /**
@@ -459,23 +462,23 @@ public class HonoKafkaConsumer implements Lifecycle {
     }
 
     private void onPartitionsAssigned(final Set<TopicPartition> partitionsSet) {
-        // if a topic pattern is used, kafkaConsumer.subscription() will be used to fetch the actual subscribed topics;
-        // if a fixed topic list is used instead, calling kafkaConsumer.subscription() would be of no use since it
-        // always returns the topics used in the kafkaConsumer.subscribe(Collection) call, regardless of whether they exist
-        // (kafkaConsumer.subscribe(Collection) would have auto-created the topics anyway, if "auto.create.topics.enable" is true)
         if (topicPattern != null) {
-            // update subscribedTopicPatternTopics (subscription() will return the actual topics, i.e. not the topic pattern if one is used
+            // update subscribedTopicPatternTopics (subscription() will return the actual topics, i.e. not the topic pattern if one is used)
             final Promise<Set<String>> subscriptionResultPromise = Promise.promise();
             kafkaConsumer.subscription(subscriptionResultPromise);
-            subscriptionResultPromise.future()
-                    .onSuccess(result -> subscribedTopicPatternTopics = new HashSet<>(result))
+            // update 'subscribedTopicPatternTopics' only after partition positions have been retrieved (ensureTopicIsAmongSubscribedTopicPatternTopics() relies on that)
+            CompositeFuture.all(subscriptionResultPromise.future(), ensurePositionsHaveBeenSetIfNeeded(partitionsSet))
+                    .onSuccess(result -> subscribedTopicPatternTopics = new HashSet<>(subscriptionResultPromise.future().result()))
                     .onFailure(thr -> log.info("failed to get subscription", thr))
-                    .map(partitionsSet)
+                    .map((Void) null)
                     .onComplete(ar -> Optional.ofNullable(subscribeDonePromiseRef.getAndSet(null))
                             .ifPresent(promise -> tryCompletePromise(promise, ar)));
         } else {
-            Optional.ofNullable(subscribeDonePromiseRef.getAndSet(null))
-                    .ifPresent(promise -> promise.tryComplete(partitionsSet));
+            // fixed topic list used: calling kafkaConsumer.subscription() here would be of no use since it always
+            // returns the topics used in the kafkaConsumer.subscribe(Collection) call, regardless of whether they exist
+            ensurePositionsHaveBeenSetIfNeeded(partitionsSet)
+                    .onComplete(ar -> Optional.ofNullable(subscribeDonePromiseRef.getAndSet(null))
+                            .ifPresent(Promise::tryComplete));
         }
         if (onPartitionsAssignedHandler != null) {
             onPartitionsAssignedHandler.handle(partitionsSet);
@@ -598,6 +601,10 @@ public class HonoKafkaConsumer implements Lifecycle {
     /**
      * Tries to ensure that the given topic is part of the list of topics this consumer is subscribed to.
      * <p>
+     * The successful completion of the returned Future means that the topic exists and is among the subscribed topics.
+     * It also means that if partitions of this topic have been assigned to this consumer, the positions for these
+     * partitions have already been fetched and the consumer will receive records published thereafter.
+     * <p>
      * Note that this method is only applicable if a topic pattern subscription is used, otherwise an
      * {@link IllegalStateException} is thrown.
      * <p>
@@ -656,13 +663,12 @@ public class HonoKafkaConsumer implements Lifecycle {
                     // therefore enforce a refresh here by again subscribing to the topic pattern
                     log.debug("ensureTopicIsAmongSubscribedTopics: verified topic existence, wait for subscription update and rebalance [{}]", topic);
                     return subscribeAndWaitForRebalance()
-                            .map((Void) null)
                             .compose(v -> {
                                 if (!subscribedTopicPatternTopics.contains(topic)) {
                                     // first metadata refresh could have failed with a LEADER_NOT_AVAILABLE error for the new topic;
                                     // seems to happen when some other topics have just been deleted for example
                                     log.debug("ensureTopicIsAmongSubscribedTopics: subscription not updated with topic after rebalance; try again [topic: {}]", topic);
-                                    return subscribeAndWaitForRebalance().mapEmpty();
+                                    return subscribeAndWaitForRebalance();
                                 }
                                 return Future.succeededFuture(v);
                             })

--- a/tests/src/test/java/org/eclipse/hono/tests/client/HonoKafkaConsumerIT.java
+++ b/tests/src/test/java/org/eclipse/hono/tests/client/HonoKafkaConsumerIT.java
@@ -23,6 +23,7 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
@@ -162,7 +163,7 @@ public class HonoKafkaConsumerIT {
             ctx.failNow(setup.causeOfFailure());
             return;
         }
-        LOG.debug("topics created and test records published");
+        LOG.debug("topics created and (to be ignored) test records published");
 
         // prepare consumer
         final Map<String, String> consumerConfig = IntegrationTestSupport.getKafkaConsumerConfig().getConsumerConfig("test");
@@ -187,6 +188,81 @@ public class HonoKafkaConsumerIT {
             LOG.debug("consumer started, publish record to be received by the consumer");
             final String recordKey = "addedAfterStartKey";
             publish(topics.iterator().next(), recordKey, Buffer.buffer("testPayload"));
+
+            nextRecordReceivedPromise.future().onComplete(ar -> {
+                ctx.verify(() -> {
+                    assertThat(receivedRecords.size()).isEqualTo(1);
+                    assertThat(receivedRecords.get(0).key()).isEqualTo(recordKey);
+                });
+                ctx.completeNow();
+            });
+        }));
+    }
+
+    /**
+     * Verifies that a HonoKafkaConsumer configured with "latest" as offset reset strategy and a topic pattern
+     * subscription only receives records published after the consumer <em>start()</em> method has completed.
+     * <p>
+     * Also verifies that all records published after the consumer <em>ensureTopicIsAmongSubscribedTopicPatternTopics()</em>
+     * method has completed are received by the consumer, also if the topic was only created after the consumer
+     * <em>start</em> method has completed.
+     *
+     * @param ctx The vert.x test context.
+     * @throws InterruptedException if test execution gets interrupted.
+     */
+    @Test
+    @Timeout(value = 10, timeUnit = TimeUnit.SECONDS)
+    public void testConsumerReadsLatestRecordsPublishedAfterTopicSubscriptionConfirmed(final VertxTestContext ctx) throws InterruptedException {
+        final String patternPrefix = "test_" + UUID.randomUUID() + "_";
+        final int numTopics = 2;
+        final Pattern topicPattern = Pattern.compile(Pattern.quote(patternPrefix) + ".*");
+        final int numPartitions = 5;
+        final int numTestRecordsPerTopic = 20;
+
+        final Set<String> topics = IntStream.range(0, numTopics)
+                .mapToObj(i -> patternPrefix + i)
+                .collect(Collectors.toSet());
+
+        final VertxTestContext setup = new VertxTestContext();
+        createTopics(topics, numPartitions)
+                .compose(v -> publishRecords(numTestRecordsPerTopic, "key_", topics))
+                .onComplete(setup.completing());
+
+        assertThat(setup.awaitCompletion(IntegrationTestSupport.getTestSetupTimeout(), TimeUnit.SECONDS)).isTrue();
+        if (setup.failed()) {
+            ctx.failNow(setup.causeOfFailure());
+            return;
+        }
+        LOG.debug("topics created and (to be ignored) test records published");
+
+        // prepare consumer
+        final Map<String, String> consumerConfig = IntegrationTestSupport.getKafkaConsumerConfig().getConsumerConfig("test");
+        consumerConfig.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest");
+
+        final AtomicReference<Promise<Void>> nextRecordReceivedPromiseRef = new AtomicReference<>();
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
+        final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler = record -> {
+            receivedRecords.add(record);
+            Optional.ofNullable(nextRecordReceivedPromiseRef.get())
+                    .ifPresent(Promise::complete);
+        };
+        kafkaConsumer = new HonoKafkaConsumer(vertx, topicPattern, recordHandler, consumerConfig);
+        // start consumer
+        kafkaConsumer.start().onComplete(ctx.succeeding(v -> {
+            ctx.verify(() -> {
+                assertThat(receivedRecords.size()).isEqualTo(0);
+            });
+            final Promise<Void> nextRecordReceivedPromise = Promise.promise();
+            nextRecordReceivedPromiseRef.set(nextRecordReceivedPromise);
+
+            LOG.debug("consumer started, create new topic implicitly by invoking ensureTopicIsAmongSubscribedTopicPatternTopics()");
+            final String newTopic = patternPrefix + "new";
+            final String recordKey = "addedAfterStartKey";
+            kafkaConsumer.ensureTopicIsAmongSubscribedTopicPatternTopics(newTopic)
+                    .onComplete(ctx.succeeding(v2 -> {
+                        LOG.debug("publish record to be received by the consumer");
+                        publish(newTopic, recordKey, Buffer.buffer("testPayload"));
+                    }));
 
             nextRecordReceivedPromise.future().onComplete(ar -> {
                 ctx.verify(() -> {

--- a/tests/src/test/java/org/eclipse/hono/tests/commandrouter/KafkaBasedCommandConsumerFactoryImplIT.java
+++ b/tests/src/test/java/org/eclipse/hono/tests/commandrouter/KafkaBasedCommandConsumerFactoryImplIT.java
@@ -21,11 +21,11 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
+import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 import javax.security.auth.x500.X500Principal;
@@ -174,13 +174,17 @@ public class KafkaBasedCommandConsumerFactoryImplIT {
         final VertxTestContext setup = new VertxTestContext();
 
         final int numTestCommands = 10;
-        final CountDownLatch allRecordsReceivedLatch = new CountDownLatch(numTestCommands);
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
+        final Promise<Void> allRecordsReceivedPromise = Promise.promise();
         final List<String> receivedCommandSubjects = new ArrayList<>();
         final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler = record -> {
+            receivedRecords.add(record);
             LOG.trace("received {}", record);
             receivedCommandSubjects.add(KafkaRecordHelper
                     .getHeaderValue(record.headers(), MessageHelper.SYS_PROPERTY_SUBJECT, String.class).orElse(""));
-            allRecordsReceivedLatch.countDown();
+            if (receivedRecords.size() == numTestCommands) {
+                allRecordsReceivedPromise.tryComplete();
+            }
         };
         final LinkedList<Promise<Void>> completionPromisesQueue = new LinkedList<>();
         // don't let getting the target adapter instance finish immediately
@@ -219,15 +223,19 @@ public class KafkaBasedCommandConsumerFactoryImplIT {
             sendOneWayCommand(tenantId, "myDeviceId", subject);
         });
 
-        if (!allRecordsReceivedLatch.await(8, TimeUnit.SECONDS)) {
-            ctx.failNow(new AssertionError(String.format("did not receive %d out of %d expected messages after 8s",
-                    allRecordsReceivedLatch.getCount(), numTestCommands)));
-        } else {
+        final long timerId = vertx.setTimer(8000, tid -> {
+            LOG.info("received records:\n{}",
+                    receivedRecords.stream().map(Object::toString).collect(Collectors.joining(",\n")));
+            allRecordsReceivedPromise.tryFail(String.format("only received %d out of %d expected messages after 8s",
+                    receivedRecords.size(), numTestCommands));
+        });
+        allRecordsReceivedPromise.future().onComplete(ctx.succeeding(v -> {
+            vertx.cancelTimer(timerId);
             ctx.verify(() -> {
                 assertThat(receivedCommandSubjects).isEqualTo(sentCommandSubjects);
             });
             ctx.completeNow();
-        }
+        }));
     }
 
     /**
@@ -261,17 +269,21 @@ public class KafkaBasedCommandConsumerFactoryImplIT {
         //   when closing the first consumer) and processes and forwards all commands starting with the second command
 
         final int numTestCommands = 10;
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
         final Promise<Void> firstRecordReceivedPromise = Promise.promise();
-        final CountDownLatch allRecordsReceivedLatch = new CountDownLatch(numTestCommands);
+        final Promise<Void> allRecordsReceivedPromise = Promise.promise();
         final List<String> receivedCommandSubjects = new ArrayList<>();
         final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler = record -> {
+            receivedRecords.add(record);
             LOG.trace("received {}", record);
             receivedCommandSubjects.add(KafkaRecordHelper
                     .getHeaderValue(record.headers(), MessageHelper.SYS_PROPERTY_SUBJECT, String.class).orElse(""));
-            if (receivedCommandSubjects.size() == 1) {
+            if (receivedRecords.size() == 1) {
                 firstRecordReceivedPromise.complete();
             }
-            allRecordsReceivedLatch.countDown();
+            if (receivedRecords.size() == numTestCommands) {
+                allRecordsReceivedPromise.tryComplete();
+            }
         };
         final Promise<Void> firstConsumerAllGetAdapterInstanceInvocationsDone = Promise.promise();
         final LinkedList<Promise<Void>> firstConsumerGetAdapterInstancePromisesQueue = new LinkedList<>();
@@ -340,17 +352,21 @@ public class KafkaBasedCommandConsumerFactoryImplIT {
                     }));
             });
 
-        if (!allRecordsReceivedLatch.await(8, TimeUnit.SECONDS)) {
-            ctx.failNow(new AssertionError(String.format("did not receive %d out of %d expected messages after 8s",
-                    allRecordsReceivedLatch.getCount(), numTestCommands)));
-        } else {
+        final long timerId = vertx.setTimer(8000, tid -> {
+            LOG.info("received records:\n{}",
+                    receivedRecords.stream().map(Object::toString).collect(Collectors.joining(",\n")));
+            allRecordsReceivedPromise.tryFail(String.format("only received %d out of %d expected messages after 8s",
+                    receivedRecords.size(), numTestCommands));
+        });
+        allRecordsReceivedPromise.future().onComplete(ctx.succeeding(v -> {
+            vertx.cancelTimer(timerId);
             ctx.verify(() -> {
                 assertThat(receivedCommandSubjects).isEqualTo(sentCommandSubjects);
                 // all but the first command should have been processed by the second consumer
                 assertThat(secondConsumerGetAdapterInstanceInvocations.get()).isEqualTo(numTestCommands - 1);
             });
             ctx.completeNow();
-        }
+        }));
     }
 
     private Future<Void> createCommandConsumer(final String tenantId,
@@ -405,6 +421,9 @@ public class KafkaBasedCommandConsumerFactoryImplIT {
     }
 
     private static void sendOneWayCommand(final String tenantId, final String deviceId, final String subject) {
+        // only sending one-way commands here since commands are possibly also received by the Command Router instance
+        // running as part of the integration test environment. For a request/response command, an error command response
+        // message would always be published there ("tenant/device not found"), interfering with any command response retrieval here
         kafkaProducer.send(getOneWayCommandRecord(tenantId, deviceId, subject), ar -> {
             if (ar.succeeded()) {
                 LOG.debug("sent command {}; metadata {}", subject, ar.result().toJson());

